publish-mqtt/src/main.rs

```diff
@@ -174,90 +173,153 @@ impl Sensor {
 
-    let bt_session = BluetoothSession::create_session(None)?;
-    let device_list = scan(&bt_session).await?;
-    let sensors = find_sensors(&bt_session, &device_list);
-    print_sensors(&sensors, &sensor_names);
-    let (named_sensors, unnamed_sensors): (Vec<_>, Vec<_>) = sensors
-        .into_iter()
-        .map(|d| Sensor::new(&d, &sensor_names).unwrap())
-        .partition(|sensor| sensor_names.contains_key(&sensor.mac_address));
-    println!(
-        "{} named sensors, {} unnamed sensors",
-        named_sensors.len(),
-        unnamed_sensors.len()
-    );
-
-    let mut sensors_to_connect: VecDeque<_> = named_sensors.into();
-
-    homie.ready().await?;
-
-    let mut sensors_connected: Vec<Sensor> = vec![];
-
-    loop {
-        connect_first_sensor_in_queue(
-            &bt_session,
-            &mut homie,
-            &mut sensors_connected,
-            &mut sensors_to_connect,
-        )
-        .await?;
+    homie
+        .ready()
+        .await
+        .with_context(|| std::line!().to_string())?;
 
-        disconnect_first_stale_sensor(&mut homie, &mut sensors_connected, &mut sensors_to_connect)
-            .await?;
+    let state = Arc::new(Mutex::new(SensorState {
+        sensors_to_connect: VecDeque::new(),
+        sensors_connected: vec![],
+        homie,
+    }));
 
-        service_bluetooth_event_queue(
-            &bt_session,
-            &mut homie,
-            &mut sensors_connected,
-            &mut sensors_to_connect,
-        )
-        .await?;
+    let mut next_scan_due = Instant::now();
+
+    let t1 = async {
+        loop {
+            let now = Instant::now();
+            if now > next_scan_due
+                && state.lock().await.sensors_connected.len() < sensor_names.len()
+            {
+                next_scan_due = now + SCAN_INTERVAL;
+                check_for_sensors(state.clone(), bt_session.clone(), &sensor_names)
+                    .await
+                    .with_context(|| std::line!().to_string())?;
+            }
+
+            {
+                let state = &mut *state.lock().await;
+                connect_first_sensor_in_queue(
+                    bt_session.clone(),
+                    &mut state.homie,
+                    &mut state.sensors_connected,
+                    &mut state.sensors_to_connect,
+                )
+                .await
+                .with_context(|| std::line!().to_string())?;
+            }
+
+            {
+                let state = &mut *state.lock().await;
+                disconnect_first_stale_sensor(
+                    &mut state.homie,
+                    &mut state.sensors_connected,
+                    &mut state.sensors_to_connect,
+                )
+                .await
+                .with_context(|| std::line!().to_string())?;
+            }
+            time::delay_for(CONNECT_INTERVAL).await;
+        }
+        #[allow(unreachable_code)]
```

Comment:
This is kind of yucky, how would it look to declare the type of t1 instead?

(We should probably also give t1 and t2 better names.)